File: src/main/kotlin/interview_150/array/InsertDeleteGetRandom.java
package interview_150.array;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Random;

/**
 * Difficulty : Medium
 * Problem Statement : https://leetcode.com/problems/insert-delete-getrandom-o1/description/?envType=study-plan-v2&envId=top-interview-150
 */

public class InsertDeleteGetRandom {

    private ArrayList<Integer> list;
    private HashMap<Integer, Integer> map;


    public InsertDeleteGetRandom() {
        list = new ArrayList<>();
        map = new HashMap<>();
    }

    private boolean search(int val) {
        return map.containsKey(val);
    }

    public boolean insert(int val) {
        if (search(val)) return false;

        list.add(val);
        map.put(val, list.size() - 1);
        return true;

    }

    public boolean remove(int val) {
        if (!search(val)) return false;

        int index = map.get(val);
        list.set(index, list.get(list.size() - 1));
        map.put(list.get(index), index);
        list.remove(list.size() - 1);
        map.remove(val);

        return true;
    }

    public int getRandom() {
        Random random = new Random();
        return list.get(random.nextInt(list.size()));
    }
}

File: src/main/kotlin/daily_challenge/InsertDeleteGetRandom.kt
package daily_challenge

/**
 * Difficulty : Medium
 * Problem Statement : https://leetcode.com/problems/insert-delete-getrandom-o1/
 */

class InsertDeleteGetRandom {

    private val set = mutableSetOf<Int>()

    fun insert(`val`: Int): Boolean {
        return set.add(`val`)
    }

    fun remove(`val`: Int): Boolean {
        return set.remove(`val`)
    }

    fun getRandom(): Int {
        return set.random()
    }

}